### PR TITLE
Add ability to import OWASP Dependency Track Finding Packaging Format (FPF) Exports as a scan

### PR DESCRIPTION
**Pull request by csansone-handy**
This PR is for issue #1560 and adds the ability to import OWASP Dependency Track Finding Packaging Format (FPF) Exports as a scan.

### CODE REVIEW

**Comment by csansone-handy** [comment]
Looking for some feedback from @stevespringett (core contributor/developer for Dependency Track) on this PR. Mainly need his thoughts on the `dojo/tools/dependency_track/parser.py` file as that is the core of where I am parsing a Dependency Track FPF file and making some (hopefully correct) assumptions on the format of each object in the findings array

File: dojo/tools/dependency_track/parser.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            return "Critical"
+        elif severity == "high":
+            return "High"
+        elif severity == "medium":
+            return "Medium"
+        elif severity == "low":
+            return "Low"
+        elif severity.startswith("info"):
+            return "Informational"
+        else:
```
Comment (stevespringett) [inline dojo/tools/dependency_track/parser.py]: Dependency-Track also supports 'UNASSIGNED' as a severity. 
https://github.com/DependencyTrack/dependency-track/blob/master/src/main/java/org/dependencytrack/model/Severity.java

This typically occurs when there's an incomplete CVE and OSS Index (or another source) picked it up. It's usually a confirmed vulnerability, but the severity is unknown as it was never formally rated.

Comment (csansone-handy) [reply dojo/tools/dependency_track/parser.py]: I believe DefectDojo only has [Critical, High, Medium, Low, Informational] so we would need to map to one of those. Here is where I default to a Critical severity if we were unable to map a Dependency Track severity to a DefectDojo severity: https://github.com/DefectDojo/django-DefectDojo/pull/1561/files#diff-0c6cbdfc684bc0d0f4300320b93fcadcR151

File: dojo/tools/dependency_track/parser.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+            raise Exception("Missing 'vulnerability' node from finding!")
+        if 'vulnId' not in dependency_track_finding['vulnerability']:
+            raise Exception("Missing 'vulnId' node from vulnerability!")
+        vuln_id = dependency_track_finding['vulnerability']['vulnId']
+        if 'source' not in dependency_track_finding['vulnerability']:
+            raise Exception("Missing 'source' node from vulnerability!")
+        source = dependency_track_finding['vulnerability']['source']
+        if 'component' not in dependency_track_finding:
+            raise Exception("Missing 'component' node from finding!")
+        if 'purl' not in dependency_track_finding['component']:
```
Comment (stevespringett) [inline dojo/tools/dependency_track/parser.py]: Not all findings will have a purl. If a component is of type (operating system or hardware device), these components will not have a purl. Software components without a purl are also likely and may involve unmanaged dependencies, dependencies that were manually added, file components.

Comment (csansone-handy) [reply dojo/tools/dependency_track/parser.py]: Hmm interesting. Good to know! Any thoughts on how we should build the title/description of a finding in DefectDojo given the data from Dependency Track. Are there attributes of the component that are guaranteed to be there (e.g. name, version, etc.)?

Comment (stevespringett) [reply dojo/tools/dependency_track/parser.py]: Component name is the only field that is required. You can use the UUID of the vulnerability to make an additional query to retrieve the full description, recommendation, credits, etc. 

The vulnerability in a finding will always include uuid, source, and vulnid, and may optionally include a title and subtitle.

Comment (csansone-handy) [reply dojo/tools/dependency_track/parser.py]: The way DefectDojo seems to work is you upload a file (either manually in the UI or via the API) that contains the scan results. This would need to be self contained and parsed for all of the needed info.

Do you have some example FPF JSON payloads you can share that show some of these things:

Component:
- No purl
- Only component name (no uid/version)

Vulnerability:
- has a title and/or subtitle

Comment (csansone-handy) [reply dojo/tools/dependency_track/parser.py]: Dependency Track seems to indicate that the `version` of a component is required (as indicated by the asterisk *).

image

As such why would the version not come back in the FPF for a Component?

Comment (stevespringett) [reply dojo/tools/dependency_track/parser.py]: Components will also have a uuid as well. That's a requirement. But version is optional.

Also note that the findings API is just one way to get the results. The vulnerability api may also be used, but you won't get the component or analysis.

FYI, ThreadFix uses the vulnerability API, Kenna Security uses the Findings API that is then converted to KDI prior to publish.

Comment (stevespringett) [reply dojo/tools/dependency_track/parser.py]: Wait hold. On. The vulnerability description and recommendation WILL be populated in the findings results if those values are non-null.

Comment (stevespringett) [reply dojo/tools/dependency_track/parser.py]: if the case of CVE, this means that a vulnerabiliy will generally have a description. In the case of NPM Audit findings, the vuln will typically have a description and recommendation.

Comment (stevespringett) [reply dojo/tools/dependency_track/parser.py]: The 'version' is not a requirement in the data model. There might be some business logic enforcement to ensure version is there, but to be safe, I would account for instances where version is null

Comment (csansone-handy) [reply dojo/tools/dependency_track/parser.py]: Okay I made a few updates to accommodate for these optional fields from Dependency Track. Let me know what you think. Diff is here: https://github.com/DefectDojo/django-DefectDojo/pull/1561/commits/ff1fb01da3d86f3f4840e30c814a4b8d1999c98f

**Comment by Maffooch** [comment]
Good wok here! The only thing I am not totally crazy about is the naming convention for each individual finding. Each one looks exactly the same at first glance and are super long.

Additionally, could you add the sample scan files [here](https://github.com/DefectDojo/sample-scan-files), and add it to list of scanners [here](https://github.com/DefectDojo/Documentation/blob/master/docs/integrations.rst). Thanks!

**Comment by stevespringett** [comment]
Looks good. I didn't see anything obvious from a Dependency-Track perspective.

**Comment by csansone-handy** [comment]
> Good wok here! The only thing I am not totally crazy about is the naming convention for each individual finding. Each one looks exactly the same at first glance and are super long.
> 
> Additionally, could you add the sample scan files [here](https://github.com/DefectDojo/sample-scan-files), and add it to list of scanners [here](https://github.com/DefectDojo/Documentation/blob/master/docs/integrations.rst). Thanks!

@Maffooch - thanks for the feedback.

Regarding the naming convention of the findings: If you look at the discussion I had with Steve above the `title` field of a vulnerability node is optional. We _could_ use the vulnerability title if it is present and make that the title of the DefectDojo finding....but then we would have cases where there is no title so we would need to fall back to what I am doing now which is making a generic title. I think I would rather have them all be consistently named.

Regarding adding the scan sample and list of scanners: will do! Thanks for pointing that out

**Comment by csansone-handy** [comment]
Added the scan sample here: https://github.com/DefectDojo/sample-scan-files/pull/37
Added this to the list of scanners here: https://github.com/DefectDojo/Documentation/pull/64

File: dojo/fixtures/test_type.json [hunk 3]
```diff
@@ -492,14 +492,20 @@
...
+	},
+	{
+		"fields": {
+			"name": "Dependency Track Finding Packaging Format (FPF) Export"
+		},
+		"model": "dojo.test_type",
+		"pk": 164
 
-	}
+	},
```
Comment (csansone-handy) [inline dojo/fixtures/test_type.json]: @Maffooch thanks for fixing my botched attempt at resolving the merge conflict. Every time I tried to pull or push it said everything was up to date so I am not sure what was going on. There is a trailing comma here that might cause some issues though!

Comment (Maffooch) [reply dojo/fixtures/test_type.json]: You're welcome! It can be a bit wonky sometimes. The trailing comma can be there or not in this case. May as well save the next person a keystroke :)

File: entrypoint_scripts/common/common-os.sh [hunk 4]
```diff
@@ -89,7 +89,7 @@ bootstrap_install() {
 	   # Install user isn't root,  check for sudo privileges
 	   echo "  Checking for sudo access, you may be prompted for your password"
        sudo -v 2>/dev/null
-       SUDO_CHECK=`sudo -v | wc -l`
+       SUDO_CHECK=`sudo -v | wc -l | awk '{$1=$1; print}'`
```
Comment (devGregA) [inline entrypoint_scripts/common/common-os.sh]: @csansone-handy was this change intended?

Comment (csansone-handy) [reply entrypoint_scripts/common/common-os.sh]: Yes it was. I was trying to run something locally (macOS Mojave) and I was having issues. Some versions of `wc -l` will emit a tab in the front so you will get `<tab>0` instead of `0`. The  `| awk '{$1=$1; print}'` just trims the whitespace

Comment (csansone-handy) [reply entrypoint_scripts/common/common-os.sh]: For example:

```
$ uname -v                                                                                                                                                                                            
Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64
$ cat ~/somefile | wc -l                                                                                                                                               
     559
```
vs.

```
$ uname -a
Linux 232842ec56f5 4.9.184-linuxkit #1 SMP Tue Jul 2 22:58:16 UTC 2019 x86_64 GNU/Linux
$ cat manage.py | wc -l
10
```
